### PR TITLE
plat-ls: Add DSPI driver for NXP LS Platforms

### PR DESCRIPTION
**Pull request by manishtomar1**
This patch adds DSPI driver for Layerscape Platforms.
DSPI compilation is enabled by default for LX2160A-QDS and LX2160A-RDB.

Signed-off-by: Carl Lamb <calamb@microsoft.com>
Signed-off-by: Manish Tomar <manish.tomar@nxp.com>

### CODE REVIEW

File: core/drivers/ls_dspi.c [hunk 1]
```diff
@@ -0,0 +1,450 @@
...
+#include <io.h>
+#include <kernel/boot.h>
+#include <kernel/delay.h>
+#include <kernel/dt.h>
+#include <libfdt.h>
+#include <mm/core_memprot.h>
+#include <platform_config.h>
+#include <util.h>
+
+#define NXP_DSPI_COMPATIBLE "fsl,lx2160a-dspi"
```
Comment (jforissier) [inline core/drivers/ls_dspi.c:20]: Minor: no need for a macro since it is used only once

File: core/drivers/ls_dspi.c [hunk 2]
```diff
@@ -0,0 +1,450 @@
+// SPDX-License-Identifier: BSD-2-Clause
+/*
+ * Copyright 2021 NXP
+ *
+ * DSPI Driver for DSPI Controller
```
Comment (jforissier) [inline core/drivers/ls_dspi.c:5]: - The first DSPI is redundant
- Maybe you could mention the meaning of DSPI here and in the commit description.

File: core/drivers/ls_dspi.c [hunk 3]
```diff
@@ -0,0 +1,450 @@
...
+ * and input clk frequency
+ */
+static TEE_Result dspi_convert_hz_to_baud(unsigned int *req_pbr,
+					  unsigned int *req_br,
+					  unsigned int speed_hz,
+					  unsigned int clkrate)
+{
+	/* Valid pre-scaler values for baud rate*/
+	static const unsigned int pbr_val[4] = { 2, 3, 5, 7 };
+
```
Comment (jforissier) [inline core/drivers/ls_dspi.c:33]: No blank line here please (only after the local variable block)

File: core/drivers/ls_dspi.c [hunk 4]
```diff
@@ -0,0 +1,450 @@
...
+	/* Valid baud rate scaler values*/
+	static const unsigned int brs_val[16] = { 2, 4, 6, 8,
+						16, 32, 64, 128,
+						256, 512, 1024, 2048,
+						4096, 8192, 16384, 32768 };
+
```
Comment (jforissier) [inline core/drivers/ls_dspi.c:39]: Same here

File: core/drivers/ls_dspi.c [hunk 5]
```diff
@@ -0,0 +1,450 @@
...
+	unsigned int tmp_val = 0, curr_val = 0;
+	unsigned int i = 0, j = 0;
```
Comment (jforissier) [inline core/drivers/ls_dspi.c:41]: Coding style is one variable per line. Several occurrences below.

File: core/drivers/ls_dspi.c [hunk 6]
```diff
@@ -0,0 +1,450 @@
...
+			curr_val = pbr_val[i] * brs_val[j];
+			if (curr_val >= tmp_val) {
+				*req_pbr = i;
+				*req_br = j;
+				return TEE_SUCCESS;
+			}
+		}
+	}
+
+	EMSG("Can not find valid baud rate,speed_hz is %d, ", speed_hz);
```
Comment (jforissier) [inline core/drivers/ls_dspi.c:56]: missing space after `,`

File: core/drivers/ls_dspi.c [hunk 7]
```diff
@@ -0,0 +1,450 @@
...
+	EMSG("clkrate is %d, we use the max prescaler value.", clkrate);
```
Comment (jforissier) [inline core/drivers/ls_dspi.c:57]: Minor:
1. "using max prescaler value" perhaps? (avoiding "we")
2. Single sentence messages usually have no trailing dot (several occurrences below as well)

File: core/drivers/ls_dspi.c [hunk 8]
```diff
@@ -0,0 +1,450 @@
...
+
+	bus_setup = io_read32(dspi_data->base + DSPI_CTAR0);
+	bus_setup &= ~(DSPI_CTAR_BRD | DSPI_CTAR_BRP(0x3) | DSPI_CTAR_BR(0xf));
+
+	status = dspi_convert_hz_to_baud(&req_i, &req_j, speed, bus_clock);
+
+	/* In case of failure scenario with max speed, setting default speed */
+	if (status == TEE_ERROR_ITEM_NOT_FOUND) {
+		speed = dspi_data->speed_hz;
+		EMSG("DSPI set_speed use default SCK rate %u.", speed);
```
Comment (jforissier) [inline core/drivers/ls_dspi.c:83]: The error message should be either here or `dspi_convert_hz_to_baud()` but probably not both?

File: core/drivers/ls_dspi.c [hunk 9]
```diff
@@ -0,0 +1,450 @@
...
+	/* wait for valid entries in RXFIFO or timeout */
+	while (DSPI_SR_RXCTR(io_read32(dspi_data->base + DSPI_SR)) == 0 &&
+	       timeout--)
+		udelay(1);
+
+	if (timeout >= 0)
+		return (uint16_t)DSPI_RFR_RXDATA(io_read32(dspi_val_addr));
+
+	EMSG("waiting timeout!");
+	return (uint16_t)(~0);
```
Comment (jforissier) [inline core/drivers/ls_dspi.c:131]: Would return 0xFFFF;` be easier to read perhaps?

File: core/drivers/ls_dspi.c [hunk 10]
```diff
@@ -0,0 +1,450 @@
...
+}
+
+static enum spi_result nxp_dspi_txrx8(struct spi_chip *chip, uint8_t *wdata,
+				      uint8_t *rdata, size_t num_pkts)
+{
+	uint8_t *spi_rd = NULL, *spi_wr = NULL;
+	uint32_t ctrl = 0;
+
+	spi_wr = wdata;
+	spi_rd = rdata;
```
Comment (jforissier) [inline core/drivers/ls_dspi.c:141]: Move instructions after declarations, same below

File: core/drivers/ls_dspi.c [hunk 11]
```diff
@@ -0,0 +1,450 @@
...
+			dspi_tx(data, ctrl, DSPI_IDLE_DATA);
+			*spi_rd++ = dspi_rx(data);
+		}
+		num_pkts = num_pkts - 1;
+	}
+
+	return SPI_OK;
+}
+
+static enum spi_result nxp_dspi_txrx16(struct spi_chip *chip, uint16_t *wdata,
```
Comment (jforissier) [inline core/drivers/ls_dspi.c:180]: This is very similar to `nxp_dspi_txrx8()`, perhaps a helper function could be used?

File: core/drivers/ls_dspi.c [hunk 12]
```diff
@@ -0,0 +1,450 @@
...
+		io_clrbits32(dspi_data->base + DSPI_MCR, DSPI_MCR_PCSIS(cs));
+	else
+		/* CSx inactive state is high */
+		io_setbits32(dspi_data->base + DSPI_MCR, DSPI_MCR_PCSIS(cs));
+}
+
+static void dspi_set_transfer_state(struct nxp_dspi_data *dspi_data,
+				    unsigned int state)
+{
+	DMSG("Set transfer state=%d bits=%d", state,
```
Comment (jforissier) [inline core/drivers/ls_dspi.c:283]: Move below declaration(s)

File: core/drivers/ls_dspi.c [hunk 13]
```diff
@@ -0,0 +1,450 @@
...
+	if (dspi_data->ctar_sel == 0)
+		io_write32(dspi_data->base + DSPI_CTAR0, bus_setup);
+	else
+		io_write32(dspi_data->base + DSPI_CTAR1, bus_setup);
+}
+
+static void dspi_set_speed(struct nxp_dspi_data *dspi_data,
+			   unsigned int speed_max_hz)
+{
+	DMSG("Set speed %d", speed_max_hz);
```
Comment (jforissier) [inline core/drivers/ls_dspi.c:314]: Redundant? (`dspi_setup_speed()` logs this information already)

File: core/drivers/ls_dspi.c [hunk 14]
```diff
@@ -0,0 +1,450 @@
...
+			if (dt_map_dev(fdt, node, &ctrl_base, &size) < 0) {
+				EMSG("Unable to get virtual address");
+				return TEE_ERROR_GENERIC;
+			}
+			break;
+		}
+	}
+
+	dspi_data->base = ctrl_base;
+
```
Comment (jforissier) [inline core/drivers/ls_dspi.c:405]: Minor: remove blank line

File: core/drivers/ls_dspi.c [hunk 15]
```diff
@@ -0,0 +1,450 @@
...
+	int node = 0;
+	vaddr_t ctrl_base = 0;
+	void *fdt = NULL;
+
+	/*
+	 * First get the DSPI Controller base address from the DTB
+	 * if DTB present and if the DSPI Controller defined in it.
+	 */
+	fdt = get_dt();
+
```
Comment (jforissier) [inline core/drivers/ls_dspi.c:380]: Minor: remove blank line

File: core/drivers/ls_dspi.c [hunk 16]
```diff
@@ -0,0 +1,450 @@
...
+static void dspi_set_transfer_state(struct nxp_dspi_data *dspi_data,
+				    unsigned int state)
+{
+	DMSG("Set transfer state=%d bits=%d", state,
+	     dspi_data->slave_data_size_bits);
+	unsigned int bus_setup = 0;
+
+	bus_setup = io_read32(dspi_data->base + DSPI_CTAR0);
+
```
Comment (jforissier) [inline core/drivers/ls_dspi.c:288]: Minor: remove blank line

File: core/drivers/ls_dspi.c [hunk 17]
```diff
@@ -0,0 +1,450 @@
...
+	DMSG("Stop DSPI Module");
+	io_setbits32(data->base + DSPI_MCR, DSPI_MCR_HALT);
+}
+
+void dspi_flush_fifo(struct nxp_dspi_data *dspi_data)
+{
+	unsigned int mcr_val = 0;
+
+	mcr_val = io_read32(dspi_data->base + DSPI_MCR);
+
```
Comment (jforissier) [inline core/drivers/ls_dspi.c:260]: Minor: remove blank line

File: core/include/drivers/ls_dspi.h [hunk 18]
```diff
@@ -0,0 +1,141 @@
...
+#define SPI_CS_HIGH   BIT(2) /* CS active high */
+#define SPI_LSB_FIRST BIT(3) /* per-word bits-on-wire */
+#define SPI_CONT      BIT(4) /* Continuous CS mode */
+
+/* default SCK frequency, unit: HZ */
+#define PLATFORM_CLK	      650000000
+#define DSPI_DEFAULT_SCK_FREQ 10000000
+#define DSPI_CLK_DIV	      4 /* prescaler divisor */
+#define DSPI_CLK	      (PLATFORM_CLK / DSPI_CLK_DIV) /* DSPI clock */
+#define CS_SPEED_MAX_HZ	      1000000	/* Slave max speed */
```
Comment (jforissier) [inline core/include/drivers/ls_dspi.h:115]: Should all these macros be defined in the header file? Why not have them in the `ls_spi.c`?

**Review by jforissier — COMMENTED** [review]
Hi @manishtomar1,

Thanks for the patch, please see my comments below. Mostly coding style or cosmetic issues, many are minor, some are even a matter of personal taste.

File: core/drivers/ls_dspi.c (hunk 11 above)
Comment (manishtomar1) [reply core/drivers/ls_dspi.c:180]: > This is very similar to `nxp_dspi_txrx8()`, perhaps a helper function could be used?

"struct spi_ops" provides support for both 8-bit and 16-bit data. So, using the same in test API to support 8-bit and 16-bit data transmission.
image

File: core/include/drivers/ls_dspi.h (hunk 18 above)
Comment (manishtomar1) [reply core/include/drivers/ls_dspi.h:115]: Move required macros to ls_dspi.c and rest are in ls_dspi.h which would be used in test API.
Not removed macros which are not being used currently considering the fact of enhancement in future.

File: core/drivers/ls_dspi.c (hunk 12 above)
Comment (jforissier) [reply core/drivers/ls_dspi.c:422]: Comment not addressed

**Comment by jforissier** [comment]
I think you missed on of my comments, with that fixed LGTM:
`Acked-by: Jerome Forissier <jerome@forissier.org>`

File: core/drivers/ls_dspi.c (hunk 12 above)
Comment (manishtomar1) [reply core/drivers/ls_dspi.c:425]: > Comment not addressed
Apologies for oversight. It is addressed now.

**Pull request merged**